types/condition.go

```diff
@@ -28,28 +31,41 @@ func getUnknownCondition(conditionType string) longhorn.Condition {
 	return condition
 }
 
-func SetConditionAndRecord(conditions map[string]longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus,
-	reason, message string, eventRecorder record.EventRecorder, obj runtime.Object, eventtype string) map[string]longhorn.Condition {
+func SetConditionAndRecord(conditions []longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus,
+	reason, message string, eventRecorder record.EventRecorder, obj runtime.Object, eventtype string) []longhorn.Condition {
 
 	condition := GetCondition(conditions, conditionType)
+
 	if condition.Status != conditionValue {
 		eventRecorder.Event(obj, eventtype, conditionType, message)
 	}
 	return SetCondition(conditions, conditionType, conditionValue, reason, message)
 }
 
-func SetCondition(originConditions map[string]longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus, reason, message string) map[string]longhorn.Condition {
-	conditions := map[string]longhorn.Condition{}
+func SetCondition(originConditions []longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus, reason, message string) []longhorn.Condition {
+	conditions := []longhorn.Condition{}
 	if originConditions != nil {
 		conditions = originConditions
 	}
+
 	condition := GetCondition(conditions, conditionType)
 	if condition.Status != conditionValue {
 		condition.LastTransitionTime = util.Now()
 	}
 	condition.Status = conditionValue
 	condition.Reason = reason
 	condition.Message = message
-	conditions[conditionType] = condition
-	return conditions
+
+	return UpdateOrAppendCondition(conditions, condition)
+}
+
+func UpdateOrAppendCondition(conditions []longhorn.Condition, condition longhorn.Condition) []longhorn.Condition {
```

Comment:
Since the `UpdateOrAppendCondition` is only used by engine_image_controller_test.go

I'd suggest considering changing the function to the private function and the public function is the `SetCondition` function. This makes there is only one public function for set conditions. Also, you could consider merging the `updateOrAppendCondition` code into `SetCondition` after the suggested change.

Reply:
Okay